File: app/src-tauri/src/commands/automation.rs
use std::{
    fs,
    path::{Path, PathBuf},
    process::Command,
    sync::Mutex,
};

use tauri::State;

use crate::{
    commands::profiles::ProfilesState,
    domain::automation::{AutomationRuntimeState, AutomationScriptEntry},
    state::automation_state::AutomationStateStore,
    utils::paths::{automation_scripts_dir, project_root},
};

pub struct AutomationState(pub Mutex<AutomationStateStore>);

#[tauri::command]
pub fn list_automation_scripts() -> Result<Vec<AutomationScriptEntry>, String> {
    let root = automation_scripts_dir();
    if !root.exists() {
        fs::create_dir_all(&root).map_err(|err| err.to_string())?;
    }

    let mut entries = fs::read_dir(&root)
        .map_err(|err| err.to_string())?
        .filter_map(Result::ok)
        .map(|entry| entry.path())
        .filter(|path| path.is_file())
        .filter(|path| {
            path.extension()
                .map(|ext| ext.to_string_lossy().to_ascii_lowercase() == "py")
                .unwrap_or(false)
        })
        .filter_map(|path| to_script_entry(&root, &path).ok())
        .collect::<Vec<_>>();

    entries.sort_by(|a, b| a.name.cmp(&b.name));
    Ok(entries)
}

#[tauri::command]
pub fn get_automation_runtime_states(
    profile_id: String,
    state: State<'_, AutomationState>,
) -> Result<Vec<AutomationRuntimeState>, String> {
    Ok(state
        .0
        .lock()
        .map_err(|err| err.to_string())?
        .get_for_profile(&profile_id))
}

#[tauri::command]
pub fn start_automation_script(
    profile_id: String,
    script_name: String,
    profiles_state: State<'_, ProfilesState>,
    state: State<'_, AutomationState>,
) -> Result<AutomationRuntimeState, String> {
    profiles_state
        .0
        .lock()
        .map_err(|err| err.to_string())?
        .get(&profile_id)?;

    validate_script_name(&script_name)?;
    let script_path = resolve_script_path(&script_name)?;

    let mut guard = state.0.lock().map_err(|err| err.to_string())?;
    guard.mark_starting(&profile_id, &script_name)?;

    let child = Command::new("python")
        .arg(script_path)
        .current_dir(project_root())
        .spawn()
        .map_err(|err| {
            let error = err.to_string();
            guard.mark_error(&profile_id, &script_name, error.clone());
            error
        })?;

    guard.mark_running(&profile_id, &script_name, child);
    Ok(guard.get(&profile_id, &script_name))
}

#[tauri::command]
pub fn stop_automation_script(
    profile_id: String,
    script_name: String,
    state: State<'_, AutomationState>,
) -> Result<AutomationRuntimeState, String> {
    validate_script_name(&script_name)?;
    Ok(state
        .0
        .lock()
        .map_err(|err| err.to_string())?
        .stop(&profile_id, &script_name))
}

#[tauri::command]
pub fn open_automation_scripts_dir() -> Result<String, String> {
    let dir = automation_scripts_dir();
    if !dir.exists() {
        fs::create_dir_all(&dir).map_err(|err| err.to_string())?;
    }

    #[cfg(target_os = "windows")]
    {
        Command::new("explorer")
            .arg(&dir)
            .spawn()
            .map_err(|err| err.to_string())?;
    }

    #[cfg(target_os = "macos")]
    {
        Command::new("open")
            .arg(&dir)
            .spawn()
            .map_err(|err| err.to_string())?;
    }

    #[cfg(all(unix, not(target_os = "macos")))]
    {
        Command::new("xdg-open")
            .arg(&dir)
            .spawn()
            .map_err(|err| err.to_string())?;
    }

    Ok(dir.to_string_lossy().to_string())
}

fn validate_script_name(script_name: &str) -> Result<(), String> {
    if script_name.trim().is_empty() {
        return Err("script name is required".into());
    }
    if script_name.contains('/') || script_name.contains('\\') || script_name.contains("..") {
        return Err("invalid script name".into());
    }
    if !script_name.to_ascii_lowercase().ends_with(".py") {
        return Err("only .py scripts are allowed".into());
    }
    Ok(())
}

fn resolve_script_path(script_name: &str) -> Result<PathBuf, String> {
    let root = automation_scripts_dir();
    if !root.exists() {
        fs::create_dir_all(&root).map_err(|err| err.to_string())?;
    }
    let candidate = root.join(script_name);
    if !candidate.exists() || !candidate.is_file() {
        return Err("script file not found".into());
    }

    let root_canonical = root.canonicalize().map_err(|err| err.to_string())?;
    let candidate_canonical = candidate.canonicalize().map_err(|err| err.to_string())?;
    if !candidate_canonical.starts_with(&root_canonical) {
        return Err("script path is outside automation directory".into());
    }

    Ok(candidate_canonical)
}

fn to_script_entry(root: &Path, path: &Path) -> Result<AutomationScriptEntry, String> {
    let name = path
        .file_name()
        .ok_or_else(|| "invalid script file name".to_string())?
        .to_string_lossy()
        .to_string();
    let relative_path = path
        .strip_prefix(root)
        .map_err(|err| err.to_string())?
        .to_string_lossy()
        .replace('\\', "/");

    Ok(AutomationScriptEntry {
        name,
        relative_path,
        absolute_path: path.to_string_lossy().to_string(),
    })
}

File: app/electron/services/automation.ts
import { spawn } from 'node:child_process'
import { mkdir, readdir } from 'node:fs/promises'
import { existsSync } from 'node:fs'
import path from 'node:path'
import type { AutomationRuntimeRecord, AutomationRuntimeState, AutomationScriptEntry } from './types.js'
import { automationScriptsDir, projectRoot } from './paths.js'
import { getProfile } from './profiles.js'

const states = new Map<string, AutomationRuntimeRecord>()

const keyOf = (profileId: string, scriptName: string) => `${profileId}::${scriptName}`

const publicState = (state: AutomationRuntimeRecord): AutomationRuntimeState => {
  const { child: _child, ...rest } = state
  return rest
}

const idleState = (profileId: string, scriptName: string): AutomationRuntimeState => ({
  profileId,
  scriptName,
  status: 'idle',
  runner: 'python',
})

const validateScriptName = (scriptName: string) => {
  if (!scriptName.trim()) {
    throw new Error('script name is required')
  }
  if (scriptName.includes('/') || scriptName.includes('\\') || scriptName.includes('..')) {
    throw new Error('invalid script name')
  }
  if (!scriptName.toLowerCase().endsWith('.py')) {
    throw new Error('only .py scripts are allowed')
  }
}

const resolveScriptPath = async (scriptName: string) => {
  validateScriptName(scriptName)
  const root = automationScriptsDir()
  await mkdir(root, { recursive: true })
  const candidate = path.join(root, scriptName)
  if (!existsSync(candidate)) {
    throw new Error('script file not found')
  }
  const rootResolved = path.resolve(root)
  const candidateResolved = path.resolve(candidate)
  if (!candidateResolved.startsWith(rootResolved)) {
    throw new Error('script path is outside automation directory')
  }
  return candidateResolved
}

export const listAutomationScripts = async (): Promise<AutomationScriptEntry[]> => {
  const root = automationScriptsDir()
  await mkdir(root, { recursive: true })
  const entries = await readdir(root, { withFileTypes: true })
  return entries
    .filter((entry) => entry.isFile() && path.extname(entry.name).toLowerCase() === '.py')
    .map((entry) => {
      const absolutePath = path.join(root, entry.name)
      return {
        name: entry.name,
        relativePath: entry.name,
        absolutePath,
      }
    })
    .sort((a, b) => a.name.localeCompare(b.name))
}

export const getAutomationRuntimeStates = async (profileId: string): Promise<AutomationRuntimeState[]> =>
  [...states.values()].filter((state) => state.profileId === profileId).map(publicState)

export const startAutomationScript = async (profileId: string, scriptName: string): Promise<AutomationRuntimeState> => {
  await getProfile(profileId)
  const scriptPath = await resolveScriptPath(scriptName)
  const key = keyOf(profileId, scriptName)
  const existing = states.get(key)
  if (existing?.child && existing.status === 'running') {
    throw new Error('script is already running')
  }

  states.set(key, {
    profileId,
    scriptName,
    status: 'starting',
    runner: 'python',
  })

  try {
    const child = spawn('python', [scriptPath], { cwd: projectRoot(), windowsHide: true })
    const pid = child.pid
    if (!pid) {
      throw new Error('failed to start automation script')
    }
    const runningState: AutomationRuntimeRecord = {
      profileId,
      scriptName,
      status: 'running',
      pid,
      startedAt: new Date().toISOString(),
      runner: 'python',
      child,
    }
    states.set(key, runningState)

    child.once('exit', (code) => {
      const latest = states.get(key)
      if (!latest || latest.pid !== pid) {
        return
      }
      states.set(key, {
        ...latest,
        child: undefined,
        status: 'stopped',
        pid: undefined,
        exitCode: code ?? undefined,
        stoppedAt: new Date().toISOString(),
      })
    })

    child.once('error', (error) => {
      states.set(key, {
        profileId,
        scriptName,
        status: 'error',
        lastError: error.message,
        stoppedAt: new Date().toISOString(),
        runner: 'python',
      })
    })

    return publicState(runningState)
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error)
    const errorState: AutomationRuntimeRecord = {
      profileId,
      scriptName,
      status: 'error',
      lastError: message,
      stoppedAt: new Date().toISOString(),
      runner: 'python',
    }
    states.set(key, errorState)
    throw new Error(message)
  }
}

export const stopAutomationScript = async (profileId: string, scriptName: string): Promise<AutomationRuntimeState> => {
  validateScriptName(scriptName)
  const key = keyOf(profileId, scriptName)
  const current: AutomationRuntimeRecord = states.get(key) ?? idleState(profileId, scriptName)
  current.child?.kill()
  const stoppedState: AutomationRuntimeRecord = {
    ...current,
    child: undefined,
    status: 'stopped',
    pid: undefined,
    stoppedAt: new Date().toISOString(),
  }
  states.set(key, stoppedState)
  return publicState(stoppedState)
}

export const openAutomationScriptsDir = async () => {
  const dir = automationScriptsDir()
  await mkdir(dir, { recursive: true })
  const { shell } = await import('electron')
  await shell.openPath(dir)
  return dir
}

export const renameAutomationProfile = (oldProfileId: string, newProfileId: string) => {
  if (oldProfileId === newProfileId) {
    return
  }
  for (const [key, state] of [...states.entries()]) {
    if (state.profileId !== oldProfileId) {
      continue
    }
    states.delete(key)
    states.set(keyOf(newProfileId, state.scriptName), { ...state, profileId: newProfileId })
  }
}
